icu4c/source/i18n/tibetancal.cpp

```diff
@@ -0,0 +1,538 @@
+// © 2016 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+/*
+ * Copyright (C) 2003-2021, International Business Machines Corporation
+ * and others. All Rights Reserved.
+ ******************************************************************************
+ *
+ * File TIBETANCAL.CPP
+ *****************************************************************************
+ */
+
+#include "tibetancal.h"
+#include <stdlib.h>
+#if !UCONFIG_NO_FORMATTING
+
+#include "umutex.h"
+#include <float.h>
+#include "gregoimp.h" // Math
+#include "uhash.h"
+
+// Debugging
+#ifdef U_DEBUG_TIBETANCAL
+#include <stdio.h>
+#include <stdarg.h>
+
+#endif
+
+U_NAMESPACE_BEGIN
+
+// Implementation of the TibetanCalendar class
+
+//-------------------------------------------------------------------------
+// Constructors...
+//-------------------------------------------------------------------------
+
+
+TibetanCalendar* TibetanCalendar::clone() const {
+  return new TibetanCalendar(*this);
+}
+
+TibetanCalendar::TibetanCalendar(const Locale& aLocale, UErrorCode& success)
+  :   Calendar(TimeZone::forLocaleOrDefault(aLocale), aLocale, success)
+{
+  setTimeInMillis(getNow(), success); // Call this again now that the vtable is set up properly.
+}
+
+TibetanCalendar::TibetanCalendar(const TibetanCalendar& other) : Calendar(other) {
+}
+
+TibetanCalendar::~TibetanCalendar()
+{
+}
+
+/**
+ * return the calendar type, "tibetan" or "tibetan-tsurphu".
+ *
+ * @return calendar type
+ * @internal
+ */
+const char *TibetanCalendar::getType() const { 
+   const char *sType = NULL;
+
+   switch (cType) {
+    case PHUGPA:
+      sType = 'tibetan';
+      break;
+    case TSURPHU:
+      sType = 'tibetan-tsurphu';
+      break;
+    default:
+      UPRV_UNREACHABLE; // out of range
+   }
+   return sType;
+}
+
+
+/**
+ * Sets Tibetan calendar calculation type used by this instance.
+ *
+ * @param type    The calendar calculation type, <code>PHUGPA</code> to use the phugpa
+ *                calendar, <code>TSURPHU</code> to use the tsurphu calendar.
+ * @internal
+ */
+void TibetanCalendar::setCalculationType(ECalculationType type, UErrorCode &status)
```

Comment:
please remove this method. We move the rest of the Calendar subclass from the hacky C style if type check way and now purely use only subclassing.